### ISSUE TITLE
训练到6个epoch出现nan

### ISSUE THREAD

**Issue by OPPOA113** [wizyoung/YOLOv3_TensorFlow#3]
![snipaste_20190123_174223](https://user-images.githubusercontent.com/4974108/51597666-b8c31b00-1f36-11e9-932f-af3fac11ac3c.jpg)
如图，训练到中间还会出现nan,什么情况。。

**Comment by OPPOA113** [comment]
忘了说明，是在自己的数据集上训练的，有6个类别。
另外，在另外一个训练集上训练的时候，还会出现如下的中断：
![snipaste_20190123_175139](https://user-images.githubusercontent.com/4974108/51598157-dc3a9580-1f37-11e9-9bbe-fd4a014a376f.jpg)
为什么会跑着跑着出现这种情况，是由于数据集的标签问题吗？程序内部没有对标签做检测吗？
帮忙看看，谢谢！！

**Comment by wizyoung** [comment]
你好，出现 nan 的情况是由于卷积网络输出的 logits 超过浮点界限导致的，这个问题的原因我**个人认为**是因为 YOLO 的输入归一化到 0-1 之间，而不是保持 0 均值，因此 bias 带来的累积效应导致前期容易输出浮点溢界。

目前无论是原作者的代码，还是第三方如 [gluon-cv](https://github.com/dmlc/gluon-cv) 等的实现都有可能遇到这个问题。他们的解决方案是加一个 warm-up 策略，你可以参考[这里](https://github.com/dmlc/gluon-cv/issues/366)或[这里](https://github.com/TencentYoutuResearch/ObjectDetection-OneStageDet/tree/master/yolo)。即前期一个小学习率先让网络适应输入，等到一定 epoches 后再让学习率复原，进行调参。带 warm-up 的方案我已经更新了。你可以先小学习率学一段时间；也可以先大学习率快速学一段时间再切回小学习率。请跟进具体实际情况调整参数。

第二个错误应该是你标签数据处理哪里错了，是不是有的图片上没有标记框，请仔细确认下。

**Comment by OPPOA113** [comment]
您好，第一个问题可以调整学习率。
但是第二个，为什么会出现这种情况，内部应该怎么对标签进行一个限制，确保标签的正确性？
另外还有一种程序跑着还会中断的情况，下图：
![snipaste_20190124_100720](https://user-images.githubusercontent.com/4974108/51649544-e1432780-1fbf-11e9-95c5-3ca373e46d5c.jpg)
这也是标签的问题吗？？

**Comment by wizyoung** [comment]
针对上面那个问题。第一个问题，可以在出错的代码处，加一个 try ... except ... 打印错误的图片的名字，核查标签是否错误；第二个，13 越界，应该是中心点坐标超出了图片范围，请自行核实下标签存储的方式是否错误。

**Comment by OPPOA113** [comment]
好的，谢谢啦。我再调试调试

**Comment by jamiechoi1995** [comment]
即使加了warm up，还是会在warm up的过程中出现NAN。。

**Comment by jamiechoi1995** [comment]
我觉得是跟bad sample有关：https://github.com/pjreddie/darknet/issues/716#issuecomment-382970887
你的代码里面https://github.com/wizyoung/YOLOv3_TensorFlow/blob/2f61e5bca30179aecdab1e9c2870df31ab1021df/utils/misc_utils.py#L31每次都会对trainlist进行shuffle，而我使用同样的参数，有时候能够收敛，有时候就出现nan。这就跟上面bad sample的猜想吻合了

**Comment by AleXander-Tsui** [comment]
@jamiechoi1995 有什么好的解决办法吗？我在COCO数据集上训练，如果不做数据增强就没有出现nan，做了数据增强就会出现。。。

**Comment by jamiechoi1995** [comment]
@AleXander-Tsui 我多尝试了几次训练，其中会有一两次不出现nan的。保存这些模型，然后后续读取这些模型用小学习率进行训练。
